### PR TITLE
storage,persist: allow callers of persist_source to handle "cannot serve..." errors

### LINKED ISSUE #19097: storage: sentry panic, [..] cannot serve requested as_of Antichain [..]
Sentry Issue: [DATABASE-BACKEND-2HR](https://materializeinc.sentry.io/issues/3893147949/?referrer=github_integration)

```
panic: u1366: s1049a9b6-bf14-4afc-ad6d-de27c16c6a46 cannot serve requested as_of Antichain { elements: [1682546362999] }: Since(Antichain { elements: [] })
  File "panicking.rs", line 575, in rust_begin_unwind
  File "shard_source.rs", line 304, in mz_persist_client::operators::shard_source::shard_source_descs::
  File "shard_source.rs", line 303, in mz_persist_client::operators::shard_source::shard_source_descs::
  File "builder_async.rs", line 500, in <mz_timely_util::builder_async::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, mz_repr::timestamp::...
  File "builder_rc.rs", line 172, in <timely::dataflow::operators::generic::builder_rc::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, m...
...
(35 additional frame(s) were not displayed)
```

ci-regexp: cannot serve requested as_of Antichain

### PR DESCRIPTION
**Pull request by aljoscha**
storage,persist: allow callers of persist_source to handle "cannot serve..." errors
    
    Partially fixes #19097
    
    Most callers will panic, as before, but they now include their purpose
    in the panic message.
    
    For storage sinks, we now trigger a graceful suspend-and-restart, which
    will get rid of at least one scenario where we previously were seeing
    this panic in CI.
    
    The machanism of the panic for sinks was:
    
    1. We create the source/sink, sending render commands to the cluster.
    2. Cluster does not start rendering yet, for reasons… delays and
       whatnot.
    3. We drop the database which immediately drops the source/sinks,
       releases critical since handles and the like.
    4. **Now the cluster wakes up and tries to render the sink. This fails
       with a panic!**
    
    With this fix, we would trigger a suspend-and-restart but eventually
    notice that the sink has been dropped and not try to re-render its
    dataflow.

Partially fixes #19097

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by jkosh44 — APPROVED** [review]

File: src/persist-client/src/operators/shard_source.rs [hunk 1]
```diff
@@ -345,7 +348,12 @@ where
             SnapshotMode::Include => match read.snapshot(as_of.clone()).await {
                 Ok(parts) => parts,
                 Err(e) => {
-                    panic!("{name_owned}: {shard_id} cannot serve requested as_of {as_of:?}: {e:?}")
+                    error_handler(format!("{name_owned}: {shard_id} cannot serve requested as_of {as_of:?}: {e:?}"));
+
+                    // We cannot continue, and we cannot shut down. Otherwise
+                    // downstream operators might interpret our downgrading our
+                    // sink as a statement of progress.
+                    future::pending().await
```
Comment (bkirwi) [inline src/persist-client/src/operators/shard_source.rs:356]: It strikes me that every call to the `error_handler` should be accompanied by a `future::pending().await`, and forgetting the second bit could cause pretty gnarly issues. Thoughts on wrapping this up behind an interface so it's easier to enforce that the two always go together?

```
struct ErrorHandler { callback: Box<dyn FnMut...> }
impl ErrorHandler { async fn report_and_stop(...) }
```

(That specific thing is probably the wrong way to do it, but you get the idea.)

**Review by bkirwi — APPROVED** [review]
Very reasonable!

File: src/persist-client/src/operators/shard_source.rs (hunk 1 above)
Comment (aljoscha) [reply src/persist-client/src/operators/shard_source.rs]: I pushed something like this and will hit merge. 👌

File: src/persist-client/src/operators/shard_source.rs [hunk 2]
```diff
@@ -221,6 +223,7 @@ pub(crate) fn shard_source_descs<K, V, D, F, G>(
     // If Some, an override for the default listen sleep retry parameters.
     listen_sleep: Option<impl Fn() -> RetryParameters + 'static>,
     start_signal: impl Future<Output = ()> + 'static,
+    mut error_handler: impl FnMut(String) -> () + 'static,
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/persist-client/src/operators/shard_source.rs]: Oh, I missed that there is one error handler that does not panic! All good, thank you

**Pull request merged**